psyneulink/core/llvm/codegen.py

```diff
@@ -21,6 +21,453 @@
+        }
+        self.name_constants = name_constants
+        super().__init__()
+
+    def visit_arguments(self, node):
+        args = node.args
+        variable = args[0]
+        # update register
+        self.register[variable.arg] = self.arg_in
+        parameters = args[1:]
+        for param in parameters:
+            assert param.arg not in ["self", "owner"], f"Unable to reference {param.arg} in a compiled UserDefinedFunction!"
+            if param.arg == 'params':
+                assert False, "Runtime parameters are not supported in compiled mode"
+            elif param.arg == 'context':
+                # Since contexts are implicit in the structs in compiled mode, we do not compile it.
+                pass
+            else:
+                self.register[param.arg] = self.func_params[param.arg]
+
+    def _get_type(self, t):
+        type_t = getattr(t, "type", t)
+        if isinstance(type_t, ir.DoubleType):
+            return ArgumentType.SCALAR
+        elif isinstance(type_t, ir.IntType) and type_t.width == 1:
+            return ArgumentType.BOOLEAN
+        elif isinstance(type_t, ir.PointerType):
+            pointee = t.type.pointee
+            return self._get_type(pointee)
+        elif isinstance(type_t, ir.ArrayType):
+            if isinstance(type_t.element, ir.ArrayType):
+                return ArgumentType.MATRIX
+            return ArgumentType.VECTOR
+        else:
+            assert False, "Don't know how to convert LLVM type: {}".format(t)
+
+    def _generate_binop(self, x, y, handlers:dict):
+        # parse args
+        x_type = self._get_type(x)
+        y_type = self._get_type(y)
+        # unpack scalars from pointers
+        if x_type == ArgumentType.SCALAR and type(getattr(x, "type", x)) is ir.PointerType:
+            x = self.builder.load(x)
+        if y_type == ArgumentType.SCALAR and type(getattr(y, "type", y)) is ir.PointerType:
+            y = self.builder.load(y)
+
+        return handlers[(x_type, y_type)](x, y)
+
+    def visit_Add(self, node):
+        def _add_vec(ctx, builder, u, v):
+            dim = len(u.type.pointee)
+            assert len(v.type.pointee) == dim
+            output_vec = builder.alloca(ir.types.ArrayType(ctx.float_ty, dim))
+
+            # Get the pointer to the first element of the array to convert from [? x double]* -> double*
+            vec_u = builder.gep(u, [ctx.int32_ty(0), ctx.int32_ty(0)])
+            vec_v = builder.gep(v, [ctx.int32_ty(0), ctx.int32_ty(0)])
+            vec_out = builder.gep(output_vec, [ctx.int32_ty(0), ctx.int32_ty(0)])
+
+            builtin = ctx.import_llvm_function("__pnl_builtin_vec_add")
+            builder.call(builtin, [vec_u, vec_v, ctx.int32_ty(dim), vec_out])
+            return output_vec
+
+        def _add_mat(ctx, builder, m1, m2):
+            x = len(m1.type.pointee)
+            y = len(m1.type.pointee.element)
+            assert len(m2.type.pointee) == x and len(m2.type.pointee.element) == y
```

Comment:
Don't use `and` in assert condition, split it into two statements.
You should be able to use `m1.type == m2.type`.